src/crimson/os/seastore/extent_placement_manager.cc

```diff
@@ -450,20 +450,14 @@ seastar::future<>
 ExtentPlacementManager::BackgroundProcess::do_background_cycle()
 {
   assert(is_ready());
-  if (trimmer->should_trim_alloc()) {
-    return trimmer->trim_alloc(
-    ).handle_error(
-      crimson::ct_error::assert_all{
-	"do_background_cycle encountered invalid error in trim_alloc"
-      }
-    );
-  } else if (trimmer->should_trim_dirty()) {
-    return trimmer->trim_dirty(
-    ).handle_error(
-      crimson::ct_error::assert_all{
-	"do_background_cycle encountered invalid error in trim_dirty"
-      }
-    );
+  if (trimmer->should_trim() &&
+      cleaner->try_reserve_projected_usage(
+          trimmer->get_trim_size_per_cycle())) {
+    return trimmer->trim(
+    ).finally([this] {
+      cleaner->release_projected_usage(
+          trimmer->get_trim_size_per_cycle());
+    });
   } else if (cleaner->should_clean_space()) {
```

Comment:
Here is one issue to be determined before merging. Should we remove this `if` condition, and just use `else` to enforce cleaner to run? Or add a state to represent the failure of the trimmer:

```cpp
  bool reserve_trimmer_success = cleaner->try_reserve_projected_usage(
          trimmer->get_trim_size_per_cycle())
  if (trimmer->should_trim() && reserve_trimmer_success) {
    return trimmer->trim(
    ).finally([this] {
      cleaner->release_projected_usage(
          trimmer->get_trim_size_per_cycle());
    });
  } else if (cleaner->should_clean_space() ||
             !reserve_trimmer_success) { // <-- make sure cleaner always run when trimmer fails reserving
    return cleaner->clean_space(
    ).handle_error() ...
```

Because I encountered this situation:
```
JournalTrimmer(
should_block_io_on_trim=0, 
should_(trim_dirty=0, trim_alloc=1), 
journal_head=jseq(sseq(1713), paddr<Seg[Dev(0),427],64946176>), 
alloc_tail=jseq(sseq(1711), paddr<Seg[Dev(0),626],64946176>), 
dirty_tail=jseq(sseq(1701), paddr<Seg[Dev(0),59],65191936>), 
alloc_tail_target=jseq(sseq(1711), paddr<Dev(BR),64946176>), 
dirty_tail_target=jseq(sseq(1701), paddr<Dev(BR),64946176>), 
tail_limit=jseq(sseq(1697), paddr<Dev(BR),64946176>)), 

SegmentCleaner(
should_block_io_on_clean=0, 
should_clean=0, 
projected_avail_ratio=0.100315, 
reclaim_ratio=0.0997815, 
alive_ratio=0.809913, 
unavailable_unreclaimable=929329152B, 
unavailable_reclaimble=47311749120B, 
alive=43427512320B, 
segments(empty=75, open=7, closed=717, type_journal=34, type_ool=690,
total=53619982336B, available=5378904064B, unavailable=48241078272B,
available_ratio=0.100315, submitted_head=jseq(sseq(1713), paddr<Seg[Dev(0),427],64950272>),
time_bound=tp(2022-11-25 10:59:15)))
```

It is caused by writing to the rbd image sequentially. Reserving will fail since projected_avail_ratio is very close to the block threshold. The cleaner also won't start because the data is sequential, and there is nothing to reclaim (reclaim_ratio=0.0997815).

This case is rare since all client io is sequential write, no overwrite. So I think this `if` condition needs removed or start cleaner directly when trimmer fails reserving space.

Reply:
>  Should we remove this if condition and just use else to force the cleaner to run? Or add a state to represent the failure of the trimmer.

I think adding a state is appropriate. It is similar in https://github.com/ceph/ceph/pull/48691#discussion_r1027644146 that `clean_cold` is forced to be true when reserving cold tier is failed.

It is not appropriate to simply drop `cleaner->should_clean_space()`, as it is responsible to delay and prevent aggressive cleaning, see https://github.com/ceph/ceph/pull/46437